server/src/main/java/keywhiz/service/daos/SecretDAO.java

```diff
@@ -217,4 +246,33 @@ public void deleteSecretByNameAndVersion(String name, String version) {
       }
     });
   }
+
+  public static class SecretDAOFactory implements DAOFactory<SecretDAO> {
+    private final DSLContext jooq;
+    private final DSLContext readonlyJooq;
+    private final SecretContentDAOFactory secretContentDAOFactory;
+    private final SecretSeriesDAOFactory secretSeriesDAOFactory;
+
+    @Inject public SecretDAOFactory(DSLContext jooq, @Readonly DSLContext readonlyJooq,
+        SecretContentDAOFactory secretContentDAOFactory,
+        SecretSeriesDAOFactory secretSeriesDAOFactory) {
+      this.jooq = jooq;
+      this.readonlyJooq = readonlyJooq;
+      this.secretContentDAOFactory = secretContentDAOFactory;
+      this.secretSeriesDAOFactory = secretSeriesDAOFactory;
+    }
+
+    @Override public SecretDAO readwrite() {
+      return new SecretDAO(jooq, secretContentDAOFactory, secretSeriesDAOFactory);
+    }
+
+    @Override public SecretDAO readonly() {
+      return new SecretDAO(readonlyJooq, secretContentDAOFactory, secretSeriesDAOFactory);
+    }
+
+    @Override public SecretDAO using(Configuration configuration) {
```

Comment:
Configuration contains settings and a DataSource (internally a Connection). Performing DSL.using(configuration) doesn't change any state, but just creates a wrapper object around the Connection. Similarly, `DSLContext.configuration()` just exposed the wrapped configuration and does not create one.

I think we should have only one of the classes (`using(Configuration)` or `using(DSLContext)`) but not both. My PR uses Configuration because that is the direct object you get from a jOOQ transaction and that's what `using` is for.

Unfortunately, the cases where we execute `dslContext.configuration()` are where we want a transaction to be used except for the MySQL/read-only/jOOQ bug. The other case is the non-public methods are being unit-tested.